File: android/app/src/main/java/StorjLib/Models/FileModel.java
package storjlib.Models;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import io.storj.libstorj.File;
import storjlib.dataProvider.contracts.FileContract;

public class FileModel {
    @Expose
    @SerializedName("bucketId")
    private String _bucketId;
    @Expose
    @SerializedName("created")
    private String _created;
    @Expose
    @SerializedName("erasure")
    private String _erasure;
    @Expose
    @SerializedName("hmac")
    private String _hmac;
    @Expose
    @SerializedName("fileId")
    private String _fileId;
    @Expose
    @SerializedName("index")
    private String _index;
    @Expose
    @SerializedName("mimeType")
    private String _mimeType;
    @Expose
    @SerializedName("name")
    private String _name;
    @Expose
    @SerializedName("size")
    private long _size;
    @Expose
    @SerializedName("isDecrypted")
    private boolean _isDecrypted;
    @Expose
    @SerializedName("isStarred")
    private boolean _isStarred;
    @Expose
    @SerializedName("isSynced")
    private boolean _isSynced;
    @Expose
    @SerializedName(FileContract._DOWNLOAD_STATE)
    private int _downloadState;
    @Expose
    @SerializedName(FileContract._FILE_HANDLE)
    private long _fileHandle;

    public FileModel(File file) {
        this(file, false, false, 0, 0);
    }

    public FileModel(File file, boolean isStarred) {
        this(file, isStarred, false, 0, 0);
    }

    public FileModel(File file, boolean isStarred, boolean isSynced) { this(file, isStarred, isSynced, 0, 0); }

    public FileModel(File file, boolean isStarred, boolean isSynced, int downloadState, long fileHandle) {
        _bucketId = file.getBucketId();
        _created = file.getCreated();
        _erasure = file.getErasure();
        _fileId = file.getId();
        _hmac = file.getHMAC();
        _index = file.getIndex();
        _isDecrypted = file.isDecrypted();
        //_mimeType = file.getMimeType();
        _name = file.getName();
        _size = file.getSize();
        _isStarred = isStarred;
        _isSynced = isSynced;
        _downloadState = downloadState;
        _fileHandle = fileHandle;
    }

    public FileModel(String bucketId, String fileId, String created, String erasure, String hmac, String index, boolean isDecrypted, boolean isStarred, String mimeType, String name, long size) {
        _bucketId = bucketId;
        _created = created;
        _erasure = erasure;
        _fileId = fileId;
        _hmac = hmac;
        _index = index;
        _isDecrypted = isDecrypted;
        _mimeType = mimeType;
        _name = name;
        _size = size;
        _isStarred = isStarred;
    }

    public FileModel(String bucketId, String fileId, String created, String erasure, String hmac, String index, boolean isDecrypted, boolean isStarred, String mimeType, String name, long size, boolean isSynced) {
        _bucketId = bucketId;
        _created = created;
        _erasure = erasure;
        _fileId = fileId;
        _hmac = hmac;
        _index = index;
        _isDecrypted = isDecrypted;
        _mimeType = mimeType;
        _name = name;
        _size = size;
        _isStarred = isStarred;
        _isSynced = isSynced;
    }

    public boolean isValid() {
        return  _fileId != null && !_fileId.isEmpty() &&
                _bucketId != null && !_bucketId.isEmpty() &&
                _name != null && !_name.isEmpty() &&
                _created != null && !_created.isEmpty() &&
                _size > 0;
    }

    public String getBucketId() {
        return _bucketId;
    }
    public String getCreated() {
        return _created;
    }
    public String getErasure() {
        return _erasure;
    }
    public String getHmac() {
        return _hmac;
    }
    public String getFileId() {
        return _fileId;
    }
    public String getIndex() {
        return _index;
    }
    public String getMimeType() {
        return _mimeType;
    }
    public String getName() {
        return _name;
    }
    public long getSize() {
        return _size;
    }
    public boolean getDecrypted() {
        return _isDecrypted;
    }
    public boolean getStarred() {
        return _isStarred;
    }
    public boolean isSynced() { return _isSynced; }
    public int downloadState() { return _downloadState; }
    public long getFileHandle() { return _fileHandle; }
}

File: ios/model/FileModel.m
//
//  FileModel.m
//  StorjMobile
//
//  Created by Barterio on 3/21/18.
//  Copyright © 2018 Storj. All rights reserved.
//

#import "FileModel.h"
#import "FileContract.h"
#import "FileDbo.h"
#import "DictionaryUtils.h"


@implementation FileModel

@synthesize _bucketId;
@synthesize _fileId;
@synthesize _created;
@synthesize _name;
@synthesize _hmac;
@synthesize _index;
@synthesize _erasure;
@synthesize _mimeType;
@synthesize _size;
@synthesize _isDecrypted;
@synthesize _isStarred;
@synthesize _isSynced;
 @synthesize _downloadState;
@synthesize _fileHandle;
@synthesize _fileUri;
@synthesize _thumbnail;

-(instancetype)initWithBucketId:(NSString *)bucketId
                        created:(NSString *)created
                        erasure:(NSString *)erasure
                           hmac:(NSString *)hmac
                         fileId:(NSString *)fileId
                          index:(NSString *)index
                       mimeType:(NSString *)mimeType
                           name:(NSString *)name
                           size:(long)size
                    isDecrypted:(BOOL)isDecrypted
                      isStarred:(BOOL)isStarred {
  
  return [self initWithBucketId:bucketId
                        created:created
                        erasure:erasure
                           hmac:hmac
                         fileId:fileId
                          index:index
                       mimeType:mimeType
                           name:name
                           size:size
                    isDecrypted:isDecrypted
                      isStarred:isStarred
                       isSynced:NO];
}

-(instancetype) initWithBucketId: (NSString *) bucketId
                         created: (NSString *) created
                         erasure: (NSString *) erasure
                            hmac: (NSString *) hmac
                          fileId: (NSString *) fileId
                           index: (NSString *) index
                        mimeType: (NSString *) mimeType
                            name: (NSString *) name
                            size: (long) size
                     isDecrypted: (BOOL) isDecrypted
                       isStarred: (BOOL) isStarred
                        isSynced: (BOOL) isSynced
                   downloadState: (int) downloadState
                      fileHandle: (long) fileHandle
                         fileUri: (NSString *) fileUri
                       thumbnail: (NSString *) thumbnail {
  if(self = [super init]) {
    _bucketId = bucketId;
    _created = created;
    _erasure = erasure;
    _hmac = hmac;
    _fileId = fileId;
    _index = index;
    _mimeType = mimeType;
    _name = name;
    _size = size;
    _isDecrypted = isDecrypted;
    _isStarred = isStarred;
    _isSynced = isSynced;
    _downloadState = downloadState;
    _fileHandle = fileHandle;
    _fileUri = fileUri;
    _thumbnail = thumbnail;
  }
  
  return self;
}

-(instancetype) initWithBucketId: (NSString *) bucketId
                         created: (NSString *) created
                         erasure: (NSString *) erasure
                            hmac: (NSString *) hmac
                          fileId: (NSString *) fileId
                           index: (NSString *) index
                        mimeType: (NSString *) mimeType
                            name: (NSString *) name
                            size: (long) size
                     isDecrypted: (BOOL) isDecrypted
                       isStarred: (BOOL) isStarred
                        isSynced: (BOOL) isSynced
                   downloadState: (int) downloadState
                         fileUri: (NSString *) fileUri
                       thumbnail: (NSString *) thumbnail {
  
  return [self initWithBucketId: _bucketId
                        created: _created
                        erasure: _erasure
                           hmac: _hmac
                         fileId: _fileId
                          index: _index
                       mimeType: _mimeType
                           name: _name
                           size: _size
                    isDecrypted: _isDecrypted
                      isStarred: _isStarred
                       isSynced: _isSynced
                  downloadState: _downloadState
                     fileHandle: _fileHandle
                        fileUri: _fileUri
                      thumbnail: _thumbnail];
}

-(instancetype) initWithSJFile: (SJFile *) sjFile {
  
  return [self initWithSJFile: sjFile
                      starred: NO
                       synced: NO
                downloadState: 0
                   fileHandle: 0
                      fileUri: nil
                    thumbnail: nil];
}

-(instancetype) initWithSJFile: (SJFile *) sjFile
                       starred: (BOOL) isStarred {
  
  return [self initWithSJFile: sjFile
                      starred: isStarred
                       synced: NO
                downloadState: 0
                   fileHandle: 0
                      fileUri: nil
                    thumbnail: nil];
}

-(instancetype) initWithSJFile: (SJFile *) sjFile
                       starred: (BOOL) isStarred
                        synced: (BOOL) isSynced {
  
  return [self initWithSJFile: sjFile
                      starred: isStarred
                       synced: isSynced
                downloadState: 0
                   fileHandle: 0
                      fileUri: nil
                    thumbnail: nil];
}

-(instancetype) initWithSJFile: (SJFile *) sjFile
                       starred: (BOOL) isStarred
                        synced: (BOOL) isSynced
                 downloadState: (int) downloadState
                    fileHandle: (long) fileHandle
                       fileUri: (NSString *) fileUri
                     thumbnail: (NSString *) thumbnail {
  
  return [self initWithBucketId: [sjFile _bucketId]
                        created: [sjFile _created]
                        erasure: [sjFile _erasure]
                           hmac: [sjFile _hmac]
                         fileId: [sjFile _fileId]
                          index: [sjFile _index]
                       mimeType: [sjFile _mimeType]
                           name: [sjFile _name]
                           size: [sjFile _size]
                    isDecrypted: [sjFile _isDecrypted]
                      isStarred: isStarred
                       isSynced: isSynced
                  downloadState: downloadState
                     fileHandle: fileHandle
                        fileUri: fileUri
                      thumbnail: thumbnail];
}

-(instancetype) initWithFileDbo: (FileDbo *) dbo {
  
  return [self initWithBucketId: [dbo _bucketId]
                        created: [dbo _created]
                        erasure: [dbo _erasure]
                           hmac: [dbo _hmac]
                         fileId: [dbo _fileId]
                          index: [dbo _index]
                       mimeType: [dbo _mimeType]
                           name: [dbo _name]
                           size: [dbo _size]
                    isDecrypted: [dbo _isDecrypted]
                      isStarred: [dbo _isStarred]
                       isSynced: [dbo _isSynced]
                  downloadState: [dbo _downloadState]
                     fileHandle: [dbo _fileHandle]
                        fileUri: [dbo _fileUri]
                      thumbnail: [dbo _thumbnail]];
}

-(BOOL) isValid {
  
  return [DictionaryUtils isNSStringValid: _bucketId]
  && [DictionaryUtils isNSStringValid: _created]
  && [DictionaryUtils isNSStringValid: _hmac]
  && [DictionaryUtils isNSStringValid: _fileId]
  && [DictionaryUtils isNSStringValid: _mimeType]
  && [DictionaryUtils isNSStringValid: _name];
}

- (NSDictionary *)toDictionary {
  NSMutableDictionary *object = [[NSMutableDictionary alloc] init];
  
  [object setObject: [DictionaryUtils checkAndReturnNSString: _bucketId]
             forKey: @FILE_MODEL_BUCKET_ID];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _created]
             forKey: FileContract.CREATED];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _erasure]
             forKey: FileContract.ERASURE];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _hmac]
             forKey: FileContract.HMAC];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _fileId]
             forKey: FileContract.FILE_ID];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _index]
             forKey:@FILE_MODEL_INDEX];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _mimeType]
             forKey: FileContract.MIME_TYPE];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _name]
             forKey: FileContract.NAME];
  [object setObject: @(_size)
             forKey: FileContract.SIZE];
  [object setObject: @(_isDecrypted)
             forKey: FileContract.DECRYPTED];
  if(_isStarred) {
    [object setObject: @(_isStarred)
               forKey: FileContract.STARRED];
  }
  [object setObject: @(_downloadState)
             forKey: FileContract.DOWNLOAD_STATE];
  [object setObject: @(_fileHandle)
             forKey: FileContract.FILE_HANDLE];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _fileUri]
             forKey: FileContract.FILE_URI];
  [object setObject: [DictionaryUtils checkAndReturnNSString: _thumbnail]
             forKey: FileContract.FILE_THUMBNAIL];
  
  return object;
}

@end
